### ISSUE TITLE
iOS _CastError exception

### ISSUE THREAD

**Issue by 1N50MN14** [ably/ably-flutter#355]
Ably periodically throws the following exception, physical device 14.8.1 I'm not sure what causes it or how to debug it, not sending any messages either. Not using the plugin to handle notifications, only messages.

{code:javascript}
he following _CastError was thrown during a platform message callback:
type 'ConnectionStateChange' is not a subtype of type 'ChannelStateChange' in type cast

When the exception was thrown, this was the stack
#0      StreamsChannel.receiveBroadcastStream.<anonymous closure>.<anonymous closure>
package:ably_flutter/…/src/streams_channel.dart:57
#1      StreamsChannel.receiveBroadcastStream.<anonymous closure>.<anonymous closure>
package:ably_flutter/…/src/streams_channel.dart:52
#2      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure>
package:flutter/…/services/binding.dart:377
#3      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure>
package:flutter/…/services/binding.dart:374
#4      _invoke2 (dart:ui/hooks.dart:187:13)
#5      _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:42:5)
#6      _Channel._drainStep (dart:ui/channel_buffers.dart:224:31)
(elided 10 frames from dart:async)
════════════════════════════════════════════════════════════════════════════════

════════ Exception caught by services library ══════════════════════════════════
type 'ChannelStateChange' is not a subtype of type 'Message' in type cast
════════════════════════════════════════════════════════════════════════════════

════════ Exception caught by services library ══════════════════════════════════
type 'ConnectionStateChange' is not a subtype of type 'ChannelStateChange' in type cast
════════════════════════════════════════════════════════════════════════════════
{code}

**Comment by ikurek** [comment]
I can't find anything obvious that can cause this, so I guess we'll have to do some more in-depth research

**Comment by ikurek** [comment]
@1N50MN14 I've spent some time with this, and I failed to find a reason that could cause this error. I'd like to get more information, so :

1. Can you tell me exactly how are you using the SDK? Maybe a code snippet of the part that causes the issue? As far as I can tell it's coming from a realtime Ably connection, but I'd like to see which method of our SDK causes this.
2. Does this occur only on this specific combination of device and os version? Can you try and see if the same thing happens on other versions of iOS or Android?

The stack trace you provided is a very generic one, coming from one of the base classes of our SDK, so it's hard to tell what the reason is without any more information

**Comment by 1N50MN14** [comment]
Apologies for the late response @ikurek had a deadline on pushing the app.

1. As far as Ably is concerned this is the only code snippet that I use:
```javascript
Future<void> reconnect() async
{
  await realtime?.close();

  realtime = null;

  final authCb = (TokenParams tokenParams) async
  {
    // just an internal auth implementation nothing special
  };

  final messageData = (Message message)
  {
      Map<String, dynamic> data = {};

    (message.data as Map).forEach((key, value)
    {
      data[key.toString()] = value;
    });

    return data;
  };

  final clientOptions = ClientOptions()
    ..clientId = accountId()
    ..logLevel = kDebugMode
      ? LogLevel.error //LogLevel.error
      : LogLevel.error
    ..autoConnect = false
    ..authMethod = 'POST'
    ..useTokenAuth = true
    ..authCallback = authCb      

  realtime = Realtime(options: clientOptions);

  await realtime!.connect();

  if(!isAnonymous())
  {
    privateDataChannel = realtime!.channels.get('private:'+accountId());

    privateDataChannel!.subscribe()..listen((message)=>
      onMessage(message.name ?? '', messageData(message)));
  }
}
```
2. I managed to replicate it, it basically happens (iOS only) whenever the app goes into background mode. If you run it either on an emulator or a real device and put the app into background mode you'll be able to replicate it.

**Comment by ikurek** [comment]
@1N50MN14 I spent some time today and I failed to reproduce the issue, here are the steps I've taken:

1. Setup iOS device emulator (iPhone 8 & iOS 15.4) 
2. Launch the `example` app from `ably-flutter`
3. Connect to realtime subscribe to channel `test-channel`
4. Publish a few messages from mobile device
5. Put the app to background (center button click on iPhone 8)
6. Send a few messages from ably.com development console
7. Restore the app to foreground (icon click on iOS launcher)

Am I missing something? The error des not appear on my device 😞  Maybe you can provide a more detailed stacktrace to identify which part of Ably library causes the issue?

**Comment by 1N50MN14** [comment]
@ikurek Apologies for the late response I've been trying to catch the error, here's how I've been able to replicate it:

1. I've set up an iOS device emulator, iPhone 13, iOS 15.4.1
2. Launched the app using the above code
3. Did nothing, publishing of messages or receiving any
4. Place the app in the background for a few seconds
5. Restored it to the foreground

This time I grabbed the full output hopefully it can give some clues:
------------------------------------------------------------------------------------------------------------
AblyFlutter.m:690 resetAblyClients()
AblyFlutter.m:690 createRealtime(AblyFlutterMessage)
AblyFlutter.m:690 connectRealtime(AblyFlutterMessage)

════════ Exception caught by services library ══════════════════════════════════
The following _CastError was thrown during a platform message callback:
type 'ConnectionStateChange' is not a subtype of type 'ChannelStateChange' in type cast

When the exception was thrown, this was the stack
#0      StreamsChannel.receiveBroadcastStream.<anonymous closure>.<anonymous closure>
package:ably_flutter/…/src/streams_channel.dart:57
#1      StreamsChannel.receiveBroadcastStream.<anonymous closure>.<anonymous closure>
package:ably_flutter/…/src/streams_channel.dart:52
#2      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure>
package:flutter/…/services/binding.dart:377
#3      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure>
package:flutter/…/services/binding.dart:374
#4      _invoke2 (dart:ui/hooks.dart:187:13)
#5      _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:42:5)
#6      _Channel._drainStep (dart:ui/channel_buffers.dart:224:31)
(elided 10 frames from dart:async)
════════════════════════════════════════════════════════════════════════════════

════════ Exception caught by services library ══════════════════════════════════
type 'ChannelStateChange' is not a subtype of type 'Message' in type cast
════════════════════════════════════════════════════════════════════════════════

════════ Exception caught by services library ══════════════════════════════════
type 'ConnectionStateChange' is not a subtype of type 'ChannelStateChange' in type cast
════════════════════════════════════════════════════════════════════════════════

**Comment by ikurek** [comment]
@1N50MN14 Thank you for the feedback, the `AblyFlutter.m:690 resetAblyClients()` stacktrace should be really helpful, as this indicates that the issue is caused by native iOS code. I'll try to reproduce the problem with the exact setup as you have, and hopefully I'll be able to identify the cause 🙂

**Comment by ikurek** [comment]
@1N50MN14 Thank you for the feedback, the `AblyFlutter.m:690 resetAblyClients()` stacktrace should be really helpful, as this indicates that the issue is caused by native iOS code. I'll try to reproduce the problem with the exact setup as you have, and hopefully I'll be able to identify the cause 🙂

**Comment by ikurek** [comment]
I've spent some time investigating (although I'm still not able to reproduce the issue) and the source seems to be this `as T` cast: 

https://github.com/ably/ably-flutter/blob/7a04c5312f24fea6c5ad624ef650d96dadf4c376/lib/src/platform/src/streams_channel.dart#L57

The `T` type however is statically determined, to I doubt there's a chance that we're trying to cast to a wrong data model, since the `T` params are declared in methods like 

https://github.com/ably/ably-flutter/blob/7a04c5312f24fea6c5ad624ef650d96dadf4c376/lib/src/platform/src/realtime/connection.dart#L67

Since this appears to be an iOS-only issue, I've checked the Obj-C part of the codebase and I think the problem may be in this method:

https://github.com/ably/ably-flutter/blob/45f6fa4dda5e2471a434cbeddb0f3d6036058da1/ios/Classes/AblyFlutterStreamHandler.m#L29

If that method emits wrong type of object on the channel, incorrect type of object will be decoded on Dart side and the type cast will fail. I'll try to investigate this further and see if I can find any cause

**Comment by 1N50MN14** [comment]
Oh I'm glad the stacktrace was helpful, yeah that makes sense as the error is iOS only.

**Comment by ikurek** [comment]
@1N50MN14 I'm still not able to reproduce the issue with our example app and Simulators, but I've discussed the problem with @lukasz-szyszkowski from iOS team and created introduced a few changes to the codebase in #399. The PR doesn't solve the underlying error (we still don't know what causes wrong types to be emitted), but it will replace the crash with an error emitted to the stream. I've also extended debug info, so we can get some more details about the type mismatch. 

I think the issue is somewhere in [ably-cocoa](https://github.com/ably/ably-cocoa), because wrong type is returned from `ARTRealtime` instance, which is a part of `ably-cocoa`, but without more details or an example where the error can be reproduced, I don't think there's much more we can do.
